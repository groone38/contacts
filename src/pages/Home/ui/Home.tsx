import React, { useState, useEffect } from "react";
import { useAppDispatch, useAppSelector } from "src/app/providers/store";
import { getContact } from "src/features/model/reducers/ContactsSlice";
import Contact from "src/entities/Contact/ui/Contact";
import classes from "./Home.module.scss";
import { Loader } from "src/shared/ui/Loader";
import { Search } from "src/features/ui/Search";

const Home = () => {
  const [search, setSearch] = useState<string>("");
  const contact = useAppSelector((state) => state.contacts.contacts);
  const loading = useAppSelector((state) => state.contacts.loading);
  const dispatch = useAppDispatch();
  useEffect(() => {
    dispatch(getContact());
  }, []);
  const filterContacts = contact.filter((item) => {
    return item.first_name
      .toLocaleLowerCase()
      .includes(search.toLocaleLowerCase());
  });
  return (
    <div className={classes.home}>
      {loading && <Loader />}
      <Search search={search} setSearch={setSearch} />
      <div className={classes.contacts}>
        {filterContacts.map((item) => (
          <Contact key={item.id} id={item.id!} name={item.first_name} />
        ))}
      </div>
    </div>
  );
};

export default Home;
