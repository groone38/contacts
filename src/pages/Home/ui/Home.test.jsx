import { act, render, screen } from "@testing-library/react";
import * as reduxHooks from "react-redux";
import "@testing-library/jest-dom";
import Home from "./Home";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";

jest.mock("react-redux");

const mockDispatch = jest.spyOn(reduxHooks, "useDispatch");
const mockSelector = jest.spyOn(reduxHooks, "useSelector");

const users = [
  { id: 1, first_name: "Andrey" },
  { id: 2, first_name: "Anton" },
];

describe("Home component", () => {
  it("render", async () => {
    mockDispatch.mockReturnValue(jest.fn());
    mockSelector.mockReturnValue([]);
    render(
      <BrowserRouter>
        <Home />
      </BrowserRouter>
    );
  });
  it("typing in search works", () => {
    mockDispatch.mockReturnValue(jest.fn());
    mockSelector.mockReturnValue(users);
    render(
      <BrowserRouter>
        <Home />
      </BrowserRouter>
    );
    expect(screen.getByText(/Andrey/i)).toBeInTheDocument();
    expect(screen.getByText(/Anton/i)).toBeInTheDocument();
    act(() => {
      userEvent.type(screen.getByRole("textbox"), "Andrey");
    });
    expect(screen.queryByDisplayValue(/Anton/)).toBeNull();
    expect(screen.queryByDisplayValue(/Andrey/)).toBeInTheDocument();
  });
  it("home snapchot", () => {
    mockDispatch.mockReturnValue(jest.fn());
    mockSelector.mockReturnValue(users);
    const home = render(
      <BrowserRouter>
        <Home />
      </BrowserRouter>
    );
    expect(home).toMatchSnapshot();
  });
});
