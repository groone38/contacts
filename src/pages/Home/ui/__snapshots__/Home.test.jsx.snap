// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component home snapchot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="home"
      >
        <div
          class="overlay"
        >
          <div
            class="lds_spinner"
          >
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
        <div
          class="search"
        >
          <input
            placeholder="Кого вы хотите найти?"
            type="text"
            value=""
          />
        </div>
        <div
          class="contacts"
        >
          <a
            class="contact"
            href="/1"
          >
            <p>
              Andrey
            </p>
          </a>
          <a
            class="contact"
            href="/2"
          >
            <p>
              Anton
            </p>
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="home"
    >
      <div
        class="overlay"
      >
        <div
          class="lds_spinner"
        >
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
      <div
        class="search"
      >
        <input
          placeholder="Кого вы хотите найти?"
          type="text"
          value=""
        />
      </div>
      <div
        class="contacts"
      >
        <a
          class="contact"
          href="/1"
        >
          <p>
            Andrey
          </p>
        </a>
        <a
          class="contact"
          href="/2"
        >
          <p>
            Anton
          </p>
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
